src/lib/chainNode.ts

```diff
@@ -48,4 +70,126 @@ export default class ChainNode {
       parent: result.parentHash.toHex(),
     }
   }
+
+  async getRoles(): Promise<RoleEnum[]> {
+    await this.api.isReady
+
+    const registry = this.api.registry
+    const lookup = registry.lookup
+    const lookupId = registry.getDefinition('DscpNodeRuntimeRole') as `Lookup${number}`
+
+    const rolesEnum = lookup.getTypeDef(lookupId).sub
+    if (Array.isArray(rolesEnum)) {
+      return rolesEnum.map((e) => ({ name: e.name, index: e.index }))
+    } else {
+      throw new Error('No roles found on-chain')
+    }
+  }
+
+  roleToIndex(role: string) {
+    const entry = this.roles.find((e) => e.name === role)
+
+    if (!entry || entry.index === undefined) {
+      throw new Error(`Invalid role: ${role}`)
+    }
+
+    return entry.index
+  }
+
+  async runProcess({ process, inputs, outputs }: Payload): Promise<number[]> {
+    await this.api.isReady
+
+    const account = this.keyring.addFromUri(USER_URI)
+
+    const outputsAsMaps = await Promise.all(
+      outputs.map(async (output: Output) => [
+        await this.processRoles(output.roles),
+        await this.processMetadata(output.metadata),
+      ])
+    )
+
+    this.logger.debug('Running Transaction inputs: %j outputs: %j', inputs, outputsAsMaps)
+
+    return new Promise((resolve, reject) => {
+      let unsub: () => void
+      this.api.tx.simpleNFT
+        .runProcess(process, inputs, outputsAsMaps)
+        .signAndSend(account, (result: SubmittableResult) => {
+          this.logger.debug('result.status %s', JSON.stringify(result.status))
+          this.logger.debug('result.status.isInBlock', result.status.isInBlock)
+          const { dispatchError, status } = result
+
+          if (dispatchError) {
+            if (dispatchError.isModule) {
+              const decoded = this.api.registry.findMetaError(dispatchError.asModule)
+              reject(new HttpResponse({ message: `Node dispatch error: ${decoded.name}` }))
+            } else {
+              reject(Error(dispatchError.toString()))
+            }
+          }
```

Comment:
what do you think of this error handling @mattdean-digicatapult, we havent used `dispatchError` before I think?

Reply:
I think we can use `dispatchError` instead of filtering the events as we have been:
```
            const errors = result.events
              .filter(({ event: { method } }) => method === 'ExtrinsicFailed')
              .map(({ event: { data } }) => data[0])
              
             if (errors.length > 0) {
              reject(ProcessExtrinsicError(errors[0]))
            }
```

e.g. for a `ProcessInvalid`, `dispatchError === errors[0]`

Reply:
I'll change L127 to a `HttpResponse` error too

Reply:
https://polkadot.js.org/docs/api/cookbook/tx/#how-do-i-get-the-decoded-enum-for-an-extrinsicfailed-event

Reply:
```
As of the @polkadot/api 2.3.1 additional result fields are exposed. Firstly there is dispatchInfo: DispatchInfo which occurs in both ExtrinsicSuccess & ExtrinsicFailed events. Additionally, on failures the dispatchError: DispatchError is exposed. With this in mind, the above can be simplified to be -
```